opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The OpenTelemetry tracing API describes the classes used to generate
+distributed traces.
+
+The :class:`.Tracer` class controls access to the execution context, and
+manages span creation. Each operation in a trace is represented by a
+:class:`.Span`, which records the start, end time, and metadata associated with
+the operation.
+
+This module provides abstract (i.e. unimplemented) classes required for
+tracing, and a concrete no-op ``BlankSpan`` that allows applications to use the
+API package alone without a supporting implementation.
+
+The tracer supports implicit and explicit context propagation. By default spans
+are created as children of the currently active span, and the newly-created
+span becomes the new active span::
+
+    from opentelemetry.sdk.trace import tracer
```

Comment:
Not sure a reference from the API to the SDK is needed here (even if it's only the docs ;) )

Reply:
How would you write the example without referencing the SDK?

Reply:
I did not notice this was the sdk package, but I agree with @carlosalberto that this bad. I think that user code should only statically reference the SDK when configuring it at startup. Regular instrumentation code like this should be written against the API only. IMHO the SDK should not even provide a global tracer object like this, because that makes it too easy to introduce a hard dependency on the SDK. The goal should be to be able to switch out the API implementation by just changing a single place in the application (configuration).
If we had a global tracer at API level (I think we should), then this could look like:

```python
from opentelementry.api import implementation as otel
with otel().tracer.span("parent"): # ...
```

Reply:
That's interesting. I expected the API to be a collection of interfaces (and a few unobjectionable concrete classes like `BlankSpan`) and for users to swap out the sdk at import time, so instead of:

```python
from opentelemetry.sdk.trace import tracer
```

they'd do:

```python
from some.other.implementation import tracer
```

But that doesn't support static config, and would require the user to change a bunch of imports to swap out the implementation. I think you're right that it should be possible to do this in one place.

Even still, using the API package as a container for the implementation is a bit weird.

@reyang do you have an opinion on this from working on the OC static config code?

Reply:
I was "inspired" by the Java implementation, which also stores references to the selected implementation in a singleton class in the API package: https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/OpenTelemetry.java#L49-L58

Reply:
One more important reason why the dependency on the SDK should not be in the instrumentation code is libraries. If a library uses the SDK, any apps and other libs that use it are more or less forced to use it too. That would defeat the purpose of having a separate API layer entirely.

I think the API package is fine as implementation container, but we can also introduce another package. Let's consider the scenario where we have an application that uses instrumentation and a library used by the app that too uses instrumentation. As implementation for the API, it uses some package imp (which could be the OpenTelemetry sdk for example). If it were in the API, we would have this dependency tree (excuse the crude ASCII-Art):
```
|app| -> |lib| -> |API|
|   | ----------> |   |
|   | -> |imp| -> |   |
```
That is, app depends on lib, API and imp, and lib and imp depend only on the API.
The dependency of app on imp is because it needs to configure it; using an Java-like approach with system properties/environment variables/config files containing class names/module names, that dependency could even be removed too/moved one layer up to the application configuration. 

With an additional implementation holder (called hld below) we would have basically the same graph but with an additional dependency of app and lib (but probably not imp) on hld:
```
|   | -------------------> |   |
|   |    |   | ----------> |   |
|app| -> |lib| -> |hld| -> |API|
|   | ----------> |   |    |   |
|   | -> |imp| ----------> |   |
```

Anyway, what I definitely want to avoid is a dependency of any lib on any particular imp (e.g., the SDK).

Reply:
This is getting quite long, I created an issue for this discussion: #15

Reply:
Thanks for the detailed explanation. I changed this docs to "api", but let's figure out exactly how loading the implementation should work in #15.